frame/contracts/src/wasm/runtime.rs

```diff
@@ -1013,8 +1022,192 @@ define_env!(Env, <E: Ext>,
+	// - The `input` and `output` buffer may overlap.
+	//
+	// # Parameters
+	//
+	// - `input_ptr`: the pointer into the linear memory where the input
+	//                data is placed.
+	// - `input_len`: the length of the input data in bytes.
+	// - `output_ptr`: the pointer into the linear memory where the output
+	//                 data is placed. The function will write the result
+	//                 directly into this buffer.
+	ext_hash_twox_256(ctx, input_ptr: u32, input_len: u32, output_ptr: u32) => {
+		compute_hash_on_scratch_buffer(ctx, twox_256, input_ptr, input_len, output_ptr)
+	},
+
+	// Computes the TWOX 128-bit hash on the given input buffer.
+	//
+	// Returns the result directly into the given output buffer.
+	//
+	// # Note
+	//
+	// - The current contents of the scratch buffer are overwritten.
+	// - The `input` and `output` buffer may overlap.
+	//
+	// # Parameters
+	//
+	// - `input_ptr`: the pointer into the linear memory where the input
+	//                data is placed.
+	// - `input_len`: the length of the input data in bytes.
+	// - `output_ptr`: the pointer into the linear memory where the output
+	//                 data is placed. The function will write the result
+	//                 directly into this buffer.
+	ext_hash_twox_128(ctx, input_ptr: u32, input_len: u32, output_ptr: u32) => {
+		compute_hash_on_scratch_buffer(ctx, twox_128, input_ptr, input_len, output_ptr)
+	},
+
+	// Computes the TWOX 64-bit hash on the given input buffer.
+	//
+	// Returns the result directly into the given output buffer.
+	//
+	// # Note
+	//
+	// - The current contents of the scratch buffer are overwritten.
+	// - The `input` and `output` buffer may overlap.
+	//
+	// # Parameters
+	//
+	// - `input_ptr`: the pointer into the linear memory where the input
+	//                data is placed.
+	// - `input_len`: the length of the input data in bytes.
+	// - `output_ptr`: the pointer into the linear memory where the output
+	//                 data is placed. The function will write the result
+	//                 directly into this buffer.
+	ext_hash_twox_64(ctx, input_ptr: u32, input_len: u32, output_ptr: u32) => {
+		compute_hash_on_scratch_buffer(ctx, twox_64, input_ptr, input_len, output_ptr)
+	},
 );
 
+/// Computes the given hash function on the scratch buffer.
+///
+/// Reads from the sandboxed input buffer into the scratch buffer.
+/// Returns the result directly to the sandboxed output buffer.
+///
+/// # Note
+///
+/// The `input` and `output` buffers may overlap.
+fn compute_hash_on_scratch_buffer<E, F, R>(
+	ctx: &mut Runtime<E>,
+	hash_fn: F,
+	input_ptr: u32,
+	input_len: u32,
+	output_ptr: u32,
+) -> Result<(), sp_sandbox::HostError>
+where
+	E: Ext,
+	F: FnOnce(&[u8]) -> R,
+	R: AsRef<[u8]> + 'static,
+{
+	// Copy the input buffer directly into the scratch buffer to avoid
+	// heap allocations.
+	read_sandbox_memory_into_scratch(ctx, input_ptr, input_len)?;
```

Comment:
I do not like this scratch buffer at all as it is something that shares data all across pallet and even with the sandboxed contracts. I consider it dangerous. And how exactly does it avoid heap allocations? It is just a Vector that resizes accordingly which can very well allocate. Yeah sometimes it won't sure. But if you think about it what we are doing is just using a `Vec` as an allocator. Why not cut this indirection and just use an allocator that handles this allocation profile well?

That said, I know that your PR just follows the general conventions already in place in the pallet. So do not consider it a blocker.

Reply:
You are right, I will remove the scratch buffer usage here so that all crypto hash `ext` functions are freed from shared state with the other `ext_` functions. Good point!

Reply:
Just to note: no, I don't think this code follows the general conventions. The general convention is to use the scratch buffer where it is necessary - and we should reserve it to the cases where it is actually required.

Reply:
What are the cases where it is *required* in your opinion? A variable sized output buffer?

Reply:
```
	ext_get_runtime_storage(ctx, key_ptr: u32, key_len: u32) -> u32 => {
		// Steal the scratch buffer so that we hopefully save an allocation for the `key_buf`.
		read_sandbox_memory_into_scratch(ctx, key_ptr, key_len)?;
```
I am confused. So is this not following general conventions either? (You wrote this 4 months ago ,@pepyakin and I took "inspiration" for the hash functions from exactly this spot.)

Reply:
@athei yeah, broadly.
@Robbepop this is the case where it is necessary - the size of the value is not known and we cannot just write it out somewhere, that's why we use the scratch buffer here. And if we clobber the scratch buffer anyway we can reuse it for the allocation. There is high chance that it already has enough capacity.

Scratch buffer is an API detail and I am not sure if we want to affect only for the sake of avoiding an extra allocation.

To be even more clearer. I must admit that such optimizations rarely make sense if they won't allow us lowering the gas price.

Reply:
> To be even more clearer. I must admit that such optimizations rarely make sense if they won't allow us lowering the gas price.

I second that.